Disable the 4-byte Packet Information header by setting the IFF_NO_PI…

ludi317 opened the pull request:
… flag when creating tun interfaces in Linux. Now userspace reads raw IPv6 packets from the kernel without needing to strip the header. The packet information header had 1) a protocol field (assume ipv6) and 2) a truncation flag to indicate whether the read was truncated (assume irrelevant, b/c buffer sizes exceed MTU).

nsd20463 commented on tun.go:
```diff
@@ -45,11 +43,6 @@ type Packet struct {
 	Body []byte
-	// The Ethernet type of the packet. Commonly seen values are
-	// 0x8000 for IPv4 and 0x86dd for IPv6.
-	Protocol uint16
```
When this field is removed then code which uses it to decide IPv4/IPv6 has to decode the body[0] to know.
It might be nicer to keep Protocol, but have tuntap set it during Read().
Truncated is used a fewer places, and is easier to remove.

ludi317 replied:
Done

ludi317 committed:
Disable the 4-byte Packet Information header by setting the IFF_NO_PI flag when creating tun interfaces in Linux. Now userspace reads raw IPv6 packets from the kernel without needing to strip the header. The packet information header had 1) a protocol field (determine protocol instead from first nibble, the version, of packet) and 2) a truncation flag to indicate whether the read was truncated (assume irrelevant, b/c buffer sizes exceed MTU).

nsd20463 committed:
restore one more place Protocol was used

nsd20463 committed:
remove flagTruncated and run the mkdefs.sh script

to regenerate the ztypes files. We don't support detection of truncation
anymore.

nsd20463 committed:
design choice: consider packets without a complete IP datagram an error

This is slightly different from the earlier code, which only considered
it a ErrShortRead if the 4 byte header was missing. But it's what we
should be checking anyway. Callers aren't checking.

2nd design choice: trim off extra bytes after the IP payload. Our
callers aren't sneaking data after the IP payload.

nsd20463 committed:
We don't need that copy during WritePacket anymore

This was, I thought, the point of using IFF_NO_PI in the first place.
But it got forgotten in the rest of the changes.

nsd20463 commented:
Ok, @ludi317 I made some changes you get to review. The last one was what I thought was going to be the point of all this, but it got left out.

ludi317 committed:
Eliminate error case first with a return. Remaining happy path code no longer needs an if-else statement.

nsd20463 commented on tun.go:
```diff
@@ -76,50 +76,49 @@ func (t *Interface) ReadPacket(buffer []byte) (Packet, error) {
 		return Packet{}, err
 	}
 	pkt := Packet{Body: buffer[:n]}
+	if len(pkt.Body) == 0 {
```
nice. cleaner.

ludi317 commented:
I pushed one minor change for more idiomatic control flow. 

The origin story of this PR was not for optimization -- it was for correctness on Junos (FreeBSD) platforms, whose ReadPacket() / WritePacket() methods do not expect any headers. 

The longer story is that we started porting over code written for Junos FreeBSD to Evo (Linux) platform, and by dumb luck one of the engineers set the `IFF_NO_PI` flag when creating the Linux tun interface. Had we used tuntap's method of creating interfaces (where the kernel prepends the 4-byte header) with Junos's ReadPacket() method (which doesn't strip the header), our end-to-end test would have failed.

nsd20463 commented:
Oh, I thought you were doing this to speed up the cloud and APs :-)
Are we ready to merge?

ludi317 commented:
I think we're ready to merge. Though I have to wonder why all the packet body length checks when reading a packet, and only 1 when writing a packet? Not that I want to add extra length checks. 

Note that this is a breaking change (no more packet.Truncated field), so all dependent repos (mist-ap and mist-edge) should be updated after they pull in the latest version of tuntap.

nsd20463 commented:
> checks on reading, not on writing
B/c you have to check on reading b/c you can't control what the other side writes to you and you don't to mess up. It's unlikely that you're write malicious garbage unless you've got a bug.

Yes, it's a breaking change to the repos which check tuntap.Packet.Truncated, and they'll know it if/when they update. I'll edit the final commit to include that info.
